### PR TITLE
Add "rewind" feature to couch to SQL forms iteration

### PR DESCRIPTION
**Pull request by millerdev**
Add a new feature making it possible to rewind the main forms iteration state so it can be resumed from an earlier point. This makes it possible to diff forms or cases that were processed, but then the resume state was lost due to a crash.

Also fix a few errors found along the way. Notably https://github.com/dimagi/commcare-hq/commit/e25326ac29f38ccce24f23eafd4540beab526a64: errors encountered while loading resume state should no longer cause that state to be lost.

🐡 Review by commit.

### CODE REVIEW

**Review by snopoke — APPROVED** [review]
Wasn't able to review in depth but changes look good generally and code outside `couch_sql_migration` is fine.

File: corehq/apps/couch_sql_migration/rewind.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                getattr(self, method)(match)
+                break
+        else:
+            raise NotImplementedError(self.move_to)
+        migration_id = self.statedb.unique_id
+        resume_key = "%s.%s.%s" % (self.domain, self.doc_type, migration_id)
+        self.itr = ResumableFunctionIterator(resume_key, None, None, None)
+
+    def resume_rewind(self, match):
+        self.offset = None
```
Comment (kaapstorm) [inline corehq/apps/couch_sql_migration/rewind.py]: Is `self.offset` an initial position, and `self.move_to` the rewound position we want to go to? "Offset" and "move to" could be synonyms, but I'm not sure what names would be clearer.

I know that `__init__` calls `__attrs_post_init__` and `__attrs_post_init__` calls this, but it still feels to me not to be an obvious place to create and initialize attributes. Would it be clearer to add this at the start of the class? e.g.
```python
    offset = attr.ib(init=False)
```

I can't see where `Rewinder` uses it. What is it for?

**Review by kaapstorm — APPROVED** [review]
This is great! And nice tests.

**Comment by kaapstorm** [comment]
:recycle: :construction_worker_man:

File: corehq/apps/couch_sql_migration/rewind.py (hunk 1 above)
Comment (millerdev) [reply corehq/apps/couch_sql_migration/rewind.py]: This should probably be explained better in a docstring or something, although I view most of this as throw-away code (hopefully it won't be used too often). It's used to indicate the rewind scanner offset:

https://github.com/dimagi/commcare-hq/blob/9df5c98e938f20a67d900aef6ba0c42cf7035add/corehq/apps/couch_sql_migration/rewind.py#L23-L26

It is used by `case_rewind` to hold the integer offset parsed form `self.move_to` string. It's set to `None` here in `resume_rewind` so it evaluates to false, causing the forms/cases scanning feature to be disabled.

`self.move_to` is a user-supplied string. Different things happen depending on its format. For example: 
- `case-1000`: scan back through 1000 previously iterated cases
- `resume-2019-09-20T14:23:14.706428Z`: set the main forms iteration state to a previously found point/state, either the original state of the iteration prior to initiating a `case-...` rewind (this is written into the statedb at the beginning of each case rewind), or the point found and written into the statedb after a `case-...` rewind has concluded.

**Pull request merged**